packages/@ember/-internals/glimmer/lib/helpers/action.ts

```diff
@@ -347,29 +351,37 @@ function makeArgsProcessor(valuePathRef: Reference | false, actionArgsRef: Refer
   }
 }
 
+function isClosureAction(action: unknown): asserts action is string | AnyFn {
+  assert(
+    'action(ref) must be a string or function',
+    typeof action === 'string' || typeof action === 'function'
+  );
+  return;
+}
+
 function makeDynamicClosureAction(
   context: object,
   targetRef: Reference<MaybeActionHandler>,
-  actionRef: Reference<string | AnyFn>,
+  actionRef: Reference<unknown>,
   processArgs: (args: unknown[]) => unknown[],
   debugKey: string
 ) {
+  let action = valueForRef(actionRef);
+
+  isClosureAction(action);
```

Comment:
Do the assert inline: `assert` is special and gets stripped in prod build. Extracting it like this will leave behind an empty function. In this case you don't gain anything by extracting it, but if you ever want/need to do it for abstraction/reuse reasons then you have to do some careful `if (DEBUG)` dance

Reply:
hm -- this fights TS -- because we cant have type assertions inline
See: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgQwM6oKazgXzgMyghDgHIABDEAIywHoATDagVwHNSBuAKG4wA9I2GAE8wGOAEEAdiIBi0uAF44ACgB0m5FDaoAXClkBtALoBKZQD5DIntwA2GePwMtpAa2kQA7tLtpMWFVSEFQOABo4UXEIfDh+ZSUVUnw3AGMYYAhpUjMDAKwYVHi4YGKZeT9efAgIVX4zO1TpDKzFGrrtNgMKhQsEXhwgA

but also, wouldn't empty functions get optimized away by a minifier, same as true/false constants?

> careful if (DEBUG) dance

this would negate the type guard :sweat_smile:

Reply:
Ah yes I see the disconnect. Generally, this works, and was what I meant:

https://www.typescriptlang.org/play?ssl=20&ssc=2&pln=11&pc=1#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXygGdCQYMAKAWxGKgHMQAueQjGLVOgGnjD2CzY8zNAGtUOAO6oAlMyIkyhXv0G5UAbgBQW0JFgIOGUoihgEAJRCJSIdCAA8AFQB88AN5b48AG5QIyEzwTloAvjp60HBIaJjqvv6BAGI4MFaIzjwW8CAAHsaowMrptvbOLi7kcIjMFnLB2rrgUQgo6EL4lFCiIADCEDiEyHAAgnF45GYdzKzsnPAAPvBJsR31PjhYwI1t453dIAAiAJ6oUJRYYP2DwyBjHZN76bXWpeYOYhLSLuubwB5eFSoVgEPbwAC8CQCIBSaWsjw66RkjW8ClIFEB3gARFN4pRkCCAEYIKAsNgcOjwVIEGLtdRYriY+AYY4ABxAOEQoI6EPBkKxswpWMWSxZ7M53PifP5uw6WMByJ03i6PWuQ1GewR6kVoSAA

The issue is that `typeof x === 'function'` only narrows to `Function` and here it is claiming that we expect it to be an `AnyFn`, which is `(...args: any[]) => any`. This distinction is kind of an (IMO) in-the-weeds TypeScript quirk, and I think in this case, in this particular file, we don't really lose any TypeSafety by swapping out `AnyFn` with `Function`, arguably it's "semantically more strict/correct" since we have no idea what kind of function is in there.

In fact, if you trace it all the way to how this is used, everything bottoms out (or tops out, I guess?) on L294/295 where it actually gets assigned as a `Function` anyway.

So:


https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXygGdCQYMAKMPYLbPALnjQGtUcB3VASkdRADdSAbgBQoSLAQp0dfERJlyAWxDEoAcxCNCGGFlTqANPCqoasxizaceBYqQyET1WrlSix4aHHj6MpRCgwBAAlEERSEHQQAB4AFQA+eABvEXh4figIZC14OJEAXxFPCR9pTDcMrJyAMRwYMMR44xD4EAAPfzMnRsjo+ISE8jhERhDbOI8Aein4AGUsJQAHCCxELBBgEu8pNAq8X1R-GFQsgAkQCCXSAfJEVEZyWHUx8L7gmKsOVASueABeJK9OD9RK2fg4LDADwdJb1DDwUCBZAQBF+UinCAXK6kchPGAveDAqIfL6cX6vCIgj41PayJKAlJpeAQEAI+6MWkyNwedIzeAAOiFxWZ9wB8CUUGYIAAIgBPU5KLBgADCEBwhGQcAAgvtUPj1FwPMz+UKBcV0vIHOQAESS5j6dQEeAjd5SGA4JRIfBsBGEAAW30O8Aw-qwTlAShwNqNhVjxXKsglUtlCqgStV6s1Or1Tz1jUpbs+qFY3wp8C5eqZ6VMOgIVf+VWyIDqDXCedkjXjlvsimZ6RtQSTSmQdYARggoPAdHoDPB6s7E24bYZ+yG5dccIh60n-nv4DaZ46bfAAD6n9eb7dDyp7xs2pd4G3M7sutla-D2kBqjVakC62QOzcWMihER9PxTH9s3-XMbwYaddEdM8KzpYDOVQg5UnSNZ4HIDANxALcd1vfdD0QgwY2rdJ4FNYVqLsBQKDI2cnTg1AnF9adA04YNQ3DREQCjGNRHSAo2ggEgqPSOAMA-Yi8BE+AigKIA

Reply:
And, for reference, the annoying debug dance looks something like this:

https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&noUnusedLocals=true&noUnusedParameters=true&useUnknownInCatchVariables=true&exactOptionalPropertyTypes=true&noFallthroughCasesInSwitch=true&noImplicitOverride=true&noPropertyAccessFromIndexSignature=true#code/CYUwxgNghgTiAEYD2A7AzgF3gEQKICEBVAcQC54AjJJCEKFAbgChRJYEMBPABwQEEUnAGIp4AXngAKAHSzYAczTl6nANoBdAJTiAfPBXNW0OPABmAVxRgMAS1T60aEDAySAtiEdR5Icphg2KPIANIiowDa2qOSWANYoSADuKJrKjs4YaGEoEVGMTAW0WDZoAsIo5JJQ1nYV8HEJydpietV58CXwZSLMTDamUnhExNoA3kzwHaWCIuJSbbUxKPFJKco19p3doi3wXLxIAwv2YqfwAEQWVnnnzAC+BVcbom5QsSAAwhBIaOZwfM8qs8lismvBxpMAPSQ+AAST2AAskOZ5AisJxkQByCAQeAoEAgYD6eAAQngCOcCEkJW2JKBeU02go4Cg5ic8AAKgBlRD0TFYJwcBFQLAAAxpMxQoqm8FApkCkRAEE4DjQNnk+OAE3g0PgQiQMHgaCQHngcCgxtEkVlSE8KH5YTc3GggXJlI6ogwFPgzqgnHkMGROVC3EDFCgFGV8AAVmysBFgPasMKAG5ChD+GxgI0gDC2IJZAKojDa3VlRJ+0L9eAY8w2vEEolQUmhRIIUGIkUytBQUwIZlgVnsi1OFzwVBRmCWLLVoYkGUYKcgbUjjKSc7HF5xygIZtPG5V6blem1Rn3JhAA

Reply:
> but also, wouldn't empty functions get optimized away by a minifier, same as true/false constants?

I'm not sure! Does it work in terser?

Reply:
But:

1. It doesn't matter in this case anyway if we switch to `Function`
2. I think even if it does, we probably don't do it (and instead do those annoying `if (DEBUG)` dance) because we don't want to rely on the function happen to be empty; it's too easy for someone to accidentally see the function lying around and add more lines during a refactor, and it's not stripped anymore, where as the other more annoying pattern makes it clear that you shouldn't do it

Reply:
> I'm not sure! Does it work in terser?

Yeah, Starbeam relies heavily on optimizing away empty functions, constructors, static blocks, etc

Reply:
Changing to `Function` all the way through affects the public API of `@ember/runloop`, so I think I'll cast on line 424.

View here: 10db7f68a